### PR TITLE
Add shRNA workflow

### PR DESCRIPTION
**Pull request by romunov**
This PR adds a short hairpin RNA differential expression process. Running the process is a two-step ordeal. First, quantification files (count matrices) are produced using workflow `trim-align-quant` which produces `data:expression:bowtie2quant` objects. To run the differential expression, an .xlsx template (see test folder) should be provided and its fields are described in the shRNA differential expression process description. Result is a DESeq2 output, along with classification of short hairpins based on groups of contrasts and thresholds (all set in the template provided as input).

## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have their ``re-save`` calls.

### CODE REVIEW

File: resolwe_bio/processes/expression/bowtie2_quant.yml [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# =======================
+# Expression from bowtie2
+# =======================
+---
+
+- slug: bowtie2-quant
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/bowtie2_quant.yml]: Why not `shrna-quant`?

File: resolwe_bio/processes/expression/bowtie2_quant.yml [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  name: Quantify species from bowtie2 output
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/bowtie2_quant.yml]: Maybe better: `Quantify shRNA species`

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 1 above)
Comment (jkokosar) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: If you will change this, update also `type` accordingly.

File: resolwe_bio/processes/expression/bowtie2_quant.yml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    sequence, `AS:i:` tag value).
+  input:
+    - name: alignment
+      label: Alignment
+      type: data:alignment:bam
+      required: true
+    - name: readlengths
+      label: Species lengths threshold
+      type: basic:integer
+      # Species with read lengths below specified threshold will be removed from final output. Default is no removal.
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/bowtie2_quant.yml]: Add this to `description: |` field.

File: resolwe_bio/processes/expression/bowtie2_quant.yml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      # {% set name = alignment_fn[:-4] if alignment_fn.endswith('.bam') else alignment_fn %}
+
+      # Prepare output files and compressed output files.
+      {% set sp_mapped = name ~ '_mapped_species.txt' %} # file for mapped species
+      {% set sp_mapped_gz = sp_mapped ~ '.gz' %} # compressed file of mapped species
+      {% set count_matrix = name ~ '_count_matrix.txt' %}
+      {% set count_matrix_gz = count_matrix ~ '.gz' %}
+
+      # Extract read species counts and alignment score.
+      echo Generating mapped species lists.
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/bowtie2_quant.yml]: `echo "Generating mapped species lists."`

File: resolwe_bio/tests/processes/test_diff_expression.py [hunk 5]
```diff
@@ -163,3 +163,111 @@ def test_edger(self):
...
+            )
+
+            sample_filenames = [pf_in + f'sample-{i}.txt.gz' for i in list(range(1, 13))]
+
+            list_expressions = []
+            for i in sample_filenames:
+                list_expressions.append(self.run_process(process.slug, {'exp': i}))
+
+        inputs = {
+            'parameter_file': pf_in + parameter_file.output['file']['file'],
```
Comment (jkokosar) [inline resolwe_bio/tests/processes/test_diff_expression.py]: If `differentialexpression-shrna` is re-worked to accept the DataField on inputs, you can write just `'parameter_file': parameter_file.id` here.

File: resolwe_bio/processes/alignment/bowtie.yml [hunk 6]
```diff
@@ -331,8 +331,23 @@
...
+
+        For --local:
+         --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00
+         --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75
+         --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)
+         --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50
+
       type: basic:string
-      default: --sensitive
+      required: false
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/alignment/bowtie.yml]: Yes, `N` and `L` are turned off, and their defaults correspond to `--sensitive` defaults.

File: resolwe_bio/processes/alignment/bowtie.yml [hunk 7]
```diff
@@ -375,6 +390,63 @@
           description: >
             The maximum fragment length for valid paired-end alignments.
           default: 500
+    - name: alignment_options
+      label: Alignment options
+      group:
+        - name: N
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/alignment/bowtie.yml]: Internet and me don't know, but I did ask about parameter precedent on [bowtie2 Issue tracker page](https://github.com/BenLangmead/bowtie2/issues/242).  I will update as soon as I find anything.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            DataField(
+                data_type='expression:bowtie2quant:',
+                required=True,
+                description='Data objects of expressions from process bowtie2-quant. These inputs should match sample '
+                            'names specified in parameter file.'
+            ),
+            label='List of expression files from bowtie2quant'
+        )
+
+    class Output:
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: I don't think we have any visualizations planned for this workflow yet. Without a clear plan, I would rather forego this for now.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            'docker': {
+                'image': 'resolwebio/rnaseq:4.2.1'
+            }
+        }
+    }
+    data_name = '{{ parameter_file.file|default("?") }}'
+
+    class Input:
+        """Input fields to process ShortHairpinRNADifferentialExpression."""
+        parameter_file = FileField(
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: Non-reproducibility is the show stopper. I will implement your suggestion.

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 1 above)
Comment (romunov) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: This is a better suggestion, thanks.

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 2 above)
Comment (romunov) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: I would prefer to have bowtie2 in the description, so my suggestion would be

> Quantify shRNA species using bowtie2

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 3 above)
Comment (romunov) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: Done.

File: resolwe_bio/processes/alignment/bowtie.yml [hunk 10]
```diff
@@ -375,6 +390,63 @@
...
+          label: Number of mismatches allowed in seed alignment (N)
+          type: basic:integer
+          required: false
+          description: >
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/alignment/bowtie.yml]: Done for all occurrences.

File: resolwe_bio/processes/alignment/bowtie.yml [hunk 11]
```diff
@@ -475,6 +547,7 @@
         fi
       fi
 
+      echo "klele sn"
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/alignment/bowtie.yml]: Done.

File: docs/CHANGELOG.rst [hunk 12]
```diff
@@ -17,6 +17,9 @@ Added
   ``workflow-bbduk-star-fc-quant-paired`` processes
 - Add independent gene filtering and gene filtering based on Cook's distance
   in ``DESeq2`` differential expression process
+- Add shRNA differential expression process. This is a two-step process which
```
[the remark this answers is not in the document]
Comment (romunov) [reply docs/CHANGELOG.rst]: Done.

File: docs/CHANGELOG.rst [hunk 13]
```diff
@@ -43,6 +46,7 @@ Changed
   ``resolwebio/rnaseq:4.3.0`` Docker image
 - Use ``resolwebio/rnaseq:4.3.0`` with Subread/featureCounts version
   1.6.3 in ``feature_counts`` process
+- Expose additional parameters of `bowtie2` process
```
[the remark this answers is not in the document]
Comment (romunov) [reply docs/CHANGELOG.rst]: Done.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    process_type = 'data:shrna:differentialexpression:'
+    version = '1.0.0'
+    category = 'Differential Expression'
+    shaduling_class = SchedulingClass.BATCH
+    entity = {'type': 'sample'}
+    requirements = {
+        'expression-engine': 'jinja',
+        'executor': {
+            'docker': {
+                'image': 'resolwebio/rnaseq:4.2.1'
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: Done.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        }
+    }
+    data_name = '{{ parameter_file.file|default("?") }}'
+
+    class Input:
+        """Input fields to process ShortHairpinRNADifferentialExpression."""
+        parameter_file = FileField(
+            label='Parameter file (.xlsx)',
+            required=True,
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: Removed for all occurrences.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        # (1) Prepare data to be pulled into R for processing. Place data objects into expression_files/ folder.
+        # (2) Pass parameter file for R's shRNAde::doDE() and execute the function.
+        # (3) Prepare outputs.
+
+        dir_expressions = './expression_files'
+        os.mkdir(dir_expressions)
+
+        # (1)
+        # Move expression files and extract files.
+        sample_list = [copy(src=x['rc']['file'], dst=dir_expressions) for x in inputs.expression_data]
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: Done.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+        for file in sample_list:
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: Changed to `fl`.

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 1 above)
Comment (romunov) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: I have renamed the yml file to `shrna_quant.yml` as well as all types and slug reference(s).

File: resolwe_bio/processes/expression/bowtie2_quant.yml (hunk 4 above)
Comment (romunov) [reply resolwe_bio/processes/expression/bowtie2_quant.yml]: Done.

File: docs/CHANGELOG.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (romunov) [reply docs/CHANGELOG.rst]: Done.

File: resolwe_bio/processes/differential_expression/shRNAde.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
+import os
+import gzip
+
+from shutil import copy
+from resolwe.process import *
+
+
+class ShortHairpinRNADifferentialExpression(Process):
+    """
+    Performing differential expression on a list of objects.
```
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/differential_expression/shRNAde.py]: A template used for testing should give a good overview of what needs to be specified. https://github.com/genialis/shRNAde/blob/master/inst/extdata/template_doDE_inputs.xlsx

File: resolwe_bio/tests/processes/test_diff_expression.py (hunk 5 above)
Comment (romunov) [reply resolwe_bio/tests/processes/test_diff_expression.py]: Done.

File: resolwe_bio/processes/workflows/trim-align-quant.yml [hunk 19]
```diff
@@ -0,0 +1,167 @@
+- slug: workflow-trim-align-quant
+  name: Trim, align and quantify using a library as a reference.
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/trim-align-quant.yml]: This name lacks the shRNA reference. Think about if `shRNA primary analysis` or something similar would be better.

**Comment by romunov** [comment]
These tests run locally:
* `test_shrna_diffexp`
* `test_shrna_quant`
* `test_run_trim_align_quant`

**Pull request merged**